src/StoreApi/Schemas/CartSchema.php

```diff
@@ -404,14 +404,28 @@ protected function get_tax_lines( $cart ) {
 	 */
 	protected function get_cart_errors( $cart ) {
 		$controller    = new CartController();
-		$item_errors   = array_filter(
-			$controller->get_cart_item_errors(),
+		$cart_errors   = [];
+		$item_errors   = $controller->get_cart_item_errors();
+		$coupon_errors = $controller->get_cart_coupon_errors();
+
+		try {
+			/**
+			 * Fire action to validate cart. Functions hooking into this should throw a \RouteException.
+			 *
+			 * @param \WC_Cart $cart Cart object.
+			 */
+			do_action( '__experimental_wooocommerce_store_api_validate_cart', $cart );
```

Comment:
Do you think it's better to use exceptions here, or should we look at a WP_Error based approach?

If we use exceptions, I wonder if it would be worth passing the exception class/namespace down so you can throw something without knowing the full RouteException namespace. Not something I've tried before, thinking out loud.

Reply:
I like exceptions, but I have the impression that there's no fundamental difference between a `RouteException` and a plain old `Exception` from a third-party developer's point of view. 

In places where we catch exceptions and convert them to `WP_Error` objects, it probably makes little sense to require folks to throw a `RouteException`. 

My intent here was to be consistent with how `CartController::validate_cart_item` throws `RouteException`. There, you can see that when `validate_cart_item` is called via `CartSchema::get_cart_errors`, exceptions are converted to `WP_Errors` again, so again we could catch plain Exceptions -- if it wasn't for the fact that `CartController::validate_cart_item` is also called via `CartController::validate_cart_items`, which does not catch `RouteException`.

I can see that:

- When we populate the cart `errors` field, all exceptions are caught and converted to `WP_Errors`, and therefore a plain `Exception` would do, for both actions -- `__experimental_wooocommerce_store_api_validate_cart` and `wooocommerce_store_api_validate_cart_item`.
- However, the `validate_cart_item` method is also called when [validating order payment](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/add/cart-order-validation-hooks/src/StoreApi/Utilities/CartController.php#L374) -- and there, we expect that any `RouteException` thrown via the `wooocommerce_store_api_validate_cart_item` action will "bubble up". What I don't like about it is that order payment will halt when the first exception is thrown. Ideally, I would probably catch all exceptions thrown via `CartController::validate_cart_item`, convert them to `WP_Error`, and include all of them inside an `InvalidCartException` instead of the current `InvalidStockLevelsInCartException` (too specific) -- then catch that downstream.

I would also argue that `__experimental_wooocommerce_store_api_validate_cart` needs to fire when [validating order payment](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/796bce91d0192ec18c3b430a5683654010240111/src/StoreApi/Routes/Checkout.php#L154), too.

Initially, before submitting the PR, I had created a `validate_cart` method and included `validate_cart_items` and `validate_cart_coupons` in it, along with the `do_action( '__experimental_wooocommerce_store_api_validate_cart' )`. However, I reverted all of that work as I was worried that it would widen the scope of the PR too much.

Reply:
Hey folks,

I have been thinking about this error-handling architecture. 

There might be a case that a cart has more than one error. This case would probably require a different process than throwing a fairly generic exception like the `RouteException`, which only can include a single string as the error message. Introducing a new `InvalidCartException` that will replace the `InvalidStockLevelsInCartException`, allows us to handle it through this new experimental filter, too. 

For example, when someone uses the `__experimental_woocommerce_store_api_validate_cart`, they could gather all errors produced in the cart and add them in a single `WP_Error`. Then this error object can be thrown using this newly created `InvalidCartException`.

A generic Exception such as the `InvalidCartException` will allow us to control all errors produced in the cart by a single extension. For example, CSP may face more than one error in the same cart based on potential complex business restrictions. I think this is the best time to deal with multiple errors in a single cart state and that this consideration will further expand the maturity of the codebase. 

Lastly, It seems that the only errors considered "critical" and will stop the checking out process are all stock-related ones, which makes sense, but we could follow a more comprehensive approach.

I could make a commit to demonstrate this architecture with example usage.

Thoughts?

Reply:
@xristos3490 let's see what Mike thinks -- I also think that the code could benefit from a slight refactor in part for the reasons mentioned by Mike, but also to introduce parity between error handling for populating the `errors` field vs validating order payment. It's possible that the code here could be made a bit more dry, too.